In [1]:
import cda_client
host='https://cda.cda-dev.broadinstitute.org'
api_client = cda_client.ApiClient(configuration=cda_client.Configuration(host=host))
api_instance = cda_client.QueryApi(api_client)

In [22]:
total_rows_query = 
res = api_instance.sql_query('v3',
                             'select count(*) row_count from gdc-bq-sample.cda_mvp.v3')
total_rows = res.result[0]['row_count']
total_rows

85111

In [73]:
res = api_instance.sql_query('v3',
                             'select count(distinct id) row_count from gdc-bq-sample.cda_mvp.v3 ')
total_unique_patients = res.result[0]['row_count']
total_unique_patients

85111

In [74]:
res = api_instance.sql_query('v3',
                             '''select count(distinct p.id) row_count
    FROM gdc-bq-sample.cda_mvp.v3 p,
    unnest(ResearchSubject) as s1,
    UNNEST(s1.identifier) as i1''')
total_unique_patients_w_subject_ids = res.result[0]['row_count']
total_unique_patients_w_subject_ids

85111

In [75]:
res = api_instance.sql_query('v3',
                             '''select count(distinct p.id) row_count
    FROM gdc-bq-sample.cda_mvp.v3 p,
    unnest(ResearchSubject) as s1,
    UNNEST(s1.Diagnosis) as d1''')
total_unique_patients_w_diagnoses = res.result[0]['row_count']
total_unique_patients_w_diagnoses

84775

In [76]:
total_unique_patients - total_unique_patients_w_diagnoses

336

List their ids

In [43]:
res = api_instance.sql_query('v3',
                             '''select distinct p.id
    FROM gdc-bq-sample.cda_mvp.v3 p,
    unnest(ResearchSubject) as s1,
    UNNEST(s1.identifier) as i1
    where p.id not in (
    select distinct p2.id
    FROM gdc-bq-sample.cda_mvp.v3 p2,
    unnest(ResearchSubject) as s2,
    UNNEST(s2.Diagnosis) as d2
    )
    ''',
                            limit=100000)
unique_patients_w_subject_ids = res.result
unique_patients_w_subject_ids

[{'id': 'HTMCP-03-06-02177'},
 {'id': 'TARGET-40-NAAWHJ'},
 {'id': 'TARGET-40-NAAWEN'},
 {'id': 'TARGET-40-NAAWEA'},
 {'id': 'TARGET-40-NAAWGZ'},
 {'id': 'TARGET-40-NAAWHK'},
 {'id': 'TARGET-40-NAAWEG'},
 {'id': 'TARGET-40-NAAWDS'},
 {'id': 'TARGET-10-PARMGY'},
 {'id': 'TARGET-40-NAAWEK'},
 {'id': 'TARGET-40-NAAWFX'},
 {'id': 'TARGET-00-BM4387'},
 {'id': 'TARGET-40-NAAWGS'},
 {'id': 'TARGET-40-NAAWDY'},
 {'id': 'TARGET-40-NAAWEF'},
 {'id': 'TARGET-40-NAAWFE'},
 {'id': 'TARGET-40-NAAWDV'},
 {'id': 'TARGET-10-SJALL040058'},
 {'id': 'TARGET-40-NAAWGY'},
 {'id': 'TARGET-00-NAAENL'},
 {'id': 'TARGET-40-NAAWFG'},
 {'id': 'TARGET-40-NAAWET'},
 {'id': 'TARGET-40-NAAWDR'},
 {'id': 'TARGET-40-NAAWDM'},
 {'id': 'TARGET-10-SJBALL014594'},
 {'id': 'TARGET-40-NAAWEL'},
 {'id': 'TARGET-40-NAAWGF'},
 {'id': 'TARGET-40-NAAWFN'},
 {'id': 'TARGET-40-NAAWGT'},
 {'id': 'TARGET-40-NAAWGH'},
 {'id': 'TARGET-00-NAAEPE'},
 {'id': 'TARGET-40-NAAWDN'},
 {'id': 'TARGET-40-NAAWFD'},
 {'id': 'TARGET-10-PARAMR'},
 {

Look at one of them

In [49]:
api_instance.sql_query('v3',
                             '''select p.*
    FROM gdc-bq-sample.cda_mvp.v3 p
    where id = 'TARGET-40-NAAWGS' ''',
                      limit=100)

{'next_url': None,
 'previous_url': None,
 'query_sql': 'select p.*\n'
              '    FROM gdc-bq-sample.cda_mvp.v3 p\n'
              "    where id = 'TARGET-40-NAAWGS' ",
 'result': [{'ResearchSubject': [{'Diagnosis': [],
                                  'Specimen': [{'File': [],
                                                'age_at_collection': None,
                                                'anatomical_site': None,
                                                'associated_project': 'TARGET-OS',
                                                'derived_from_specimen': 'Initial '
                                                                         'sample',
                                                'derived_from_subject': 'TARGET-40-NAAWGS',
                                                'id': 'd0d90625-eead-5278-ba91-43ac95dcaa4e',
                                                'identifier': [{'system': 'GDC',
                                               

Get ids for some with diagnoses

In [50]:
api_instance.sql_query('v3',
                             '''select distinct p.id 
    FROM gdc-bq-sample.cda_mvp.v3 p,
    unnest(ResearchSubject) as s1,
    UNNEST(s1.Diagnosis) as d1''',
                      limit=100)

{'next_url': None,
 'previous_url': None,
 'query_sql': 'select distinct p.id \n'
              '    FROM gdc-bq-sample.cda_mvp.v3 p,\n'
              '    unnest(ResearchSubject) as s1,\n'
              '    UNNEST(s1.Diagnosis) as d1',
 'result': [{'id': 'C3L-03513'},
            {'id': 'GTEX-QVJO-0011-R10A-SM-HAKXV'},
            {'id': 'GTEX-NPJ7-0011-R10A-SM-HAKXW'},
            {'id': 'GTEX-RN5K-0011-R10A-SM-HAKXU'},
            {'id': 'GTEX-R55F-0011-R10A-SM-HAKXY'},
            {'id': 'GENIE-MSK-P-0015550'},
            {'id': 'GENIE-MSK-P-0012319'},
            {'id': 'GENIE-DFCI-003418'},
            {'id': 'GENIE-DFCI-002519'},
            {'id': 'GENIE-GRCC-134c6659'},
            {'id': 'GENIE-GRCC-e2229b2f'},
            {'id': 'GENIE-JHU-03864'},
            {'id': 'GENIE-JHU-03879'},
            {'id': 'GENIE-JHU-03582'},
            {'id': 'GENIE-JHU-00590'},
            {'id': 'GENIE-JHU-03295'},
            {'id': 'GENIE-JHU-03596'},
            {'id': 'GENIE-JHU-038

Look at one of them

In [51]:
api_instance.sql_query('v3',
                             '''select p.*
    FROM gdc-bq-sample.cda_mvp.v3 p
    where id = 'GENIE-MDA-2875' ''',
                      limit=100)

{'next_url': None,
 'previous_url': None,
 'query_sql': 'select p.*\n'
              '    FROM gdc-bq-sample.cda_mvp.v3 p\n'
              "    where id = 'GENIE-MDA-2875' ",
 'result': [{'ResearchSubject': [{'Diagnosis': [{'Treatment': [],
                                                 'age_at_diagnosis': None,
                                                 'id': 'bdf63356-3d32-4f37-b4c5-edb461ddaf5c',
                                                 'morphology': '8000/3',
                                                 'primary_diagnosis': 'Neoplasm, '
                                                                      'malignant',
                                                 'tumor_grade': 'Not Reported',
                                                 'tumor_stage': 'not '
                                                                'reported'}],
                                  'Specimen': [{'File': [{'associated_project': ['GENIE-MDA'],
                          

check count when we also nest Treatment

In [55]:
res = api_instance.sql_query('v3',
                             '''select count(distinct p.id) row_count
    FROM gdc-bq-sample.cda_mvp.v3 p,
    unnest(ResearchSubject) as s1,
    UNNEST(s1.Diagnosis) as d1,
    unnest(d1.Treatment) as t1 ''')
total_unique_patients_w_treatment = res.result[0]['row_count']
total_unique_patients_w_treatment

12560

In [57]:
def queryModV3(api_instance, query, limit=None):
    modQuery = '''
    with new_v3 as (select p.*,  d1.Treatment as treatment
    FROM gdc-bq-sample.cda_mvp.v3 p,
    unnest(ResearchSubject) as s1,
    UNNEST(s1.identifier) as i1,
    UNNEST(s1.Diagnosis) as d1
    where i1.system = 'GDC') 
    {}
    '''.format(query)
    return api_instance.sql_query('v3',modQuery, limit=10)

In [62]:
queryModV3(api_instance, 'select count(distinct pa.id) from new_v3 pa')

{'next_url': None,
 'previous_url': None,
 'query_sql': '\n'
              '    with new_v3 as (select p.*,  d1.Treatment as treatment\n'
              '    FROM gdc-bq-sample.cda_mvp.v3 p,\n'
              '    unnest(ResearchSubject) as s1,\n'
              '    UNNEST(s1.identifier) as i1,\n'
              '    UNNEST(s1.Diagnosis) as d1\n'
              "    where i1.system = 'GDC') \n"
              '    select count(distinct pa.id) from new_v3 pa\n'
              '    ',
 'result': [{'': 84152}]}

In [61]:
api_instance.sql_query('v3',
                             '''
    select count(distinct p.id)
    FROM gdc-bq-sample.cda_mvp.v3 p,
    unnest(ResearchSubject) as s1,
    UNNEST(s1.identifier) as i1,
    UNNEST(s1.Diagnosis) as d1 ''')

{'next_url': None,
 'previous_url': None,
 'query_sql': '\n'
              '    select count(distinct p.id)\n'
              '    FROM gdc-bq-sample.cda_mvp.v3 p,\n'
              '    unnest(ResearchSubject) as s1,\n'
              '    UNNEST(s1.identifier) as i1,\n'
              '    UNNEST(s1.Diagnosis) as d1 ',
 'result': [{'': 84775}]}

We don't want to filter this down to cases where there is a diagnosis.

Is there a way to do an outer join with diagnosis?

The above is essentially the same as a cross join against each unnested table. Reriting the query to include cross joins gives the same result.

In [69]:
api_instance.sql_query('v3',
                             '''
    select count(distinct p.id) res_count
    FROM gdc-bq-sample.cda_mvp.v3 p
    cross join unnest(ResearchSubject) as s1
    cross join UNNEST(s1.identifier) as i1
    cross join UNNEST(s1.Diagnosis) as d1 
    ''')

{'next_url': None,
 'previous_url': None,
 'query_sql': '\n'
              '    select count(distinct p.id) res_count\n'
              '    FROM gdc-bq-sample.cda_mvp.v3 p\n'
              '    cross join unnest(ResearchSubject) as s1\n'
              '    cross join UNNEST(s1.identifier) as i1\n'
              '    cross join UNNEST(s1.Diagnosis) as d1 \n'
              '    ',
 'result': [{'res_count': 84775}]}

Thhough it gives the same result it suggests the possibility of changing the join type. As some of the research subjects are missing diagnosis we can try a left outer join when unnesting diagnosis.

In [68]:
api_instance.sql_query('v3',
                             '''
    select count(distinct p.id) res_count
    FROM gdc-bq-sample.cda_mvp.v3 p
    cross join unnest(ResearchSubject) as s1
    cross join UNNEST(s1.identifier) as i1
    left outer join UNNEST(s1.Diagnosis) as d1 
    ''')

{'next_url': None,
 'previous_url': None,
 'query_sql': '\n'
              '    select count(distinct p.id)\n'
              '    FROM gdc-bq-sample.cda_mvp.v3 p\n'
              '    cross join unnest(ResearchSubject) as s1\n'
              '    cross join UNNEST(s1.identifier) as i1\n'
              '    left outer join UNNEST(s1.Diagnosis) as d1 \n'
              '    ',
 'result': [{'': 85111}]}

That restores the proper count of patients.

We expect that the following would give the count which accounts for the difference i.e. the 336 rows that are added by the outer join.

In [79]:
api_instance.sql_query('v3',
                             '''
    select count(distinct p.id) res_count
    FROM gdc-bq-sample.cda_mvp.v3 p
    cross join unnest(ResearchSubject) as s1
    cross join UNNEST(s1.identifier) as i1
    left outer join UNNEST(s1.Diagnosis) as d1 
    where d1 is  null

    ''')

{'next_url': None,
 'previous_url': None,
 'query_sql': '\n'
              '    select count(distinct p.id) res_count\n'
              '    FROM gdc-bq-sample.cda_mvp.v3 p\n'
              '    cross join unnest(ResearchSubject) as s1\n'
              '    cross join UNNEST(s1.identifier) as i1\n'
              '    left outer join UNNEST(s1.Diagnosis) as d1 \n'
              '    where d1 is  null\n'
              '\n'
              '    ',
 'result': [{'res_count': 398}]}

That's not the case. We'll assume for now that the difference is that there are some patients with more than one research subject one of which has a diagnosis and another that does not.

In [91]:
ids1 = [ sub['id'] for sub in unique_patients_w_subject_ids ]
len(ids1)

336

In [90]:
res2 = api_instance.sql_query('v3',
                             '''
    select distinct p.id
    FROM gdc-bq-sample.cda_mvp.v3 p
    cross join unnest(ResearchSubject) as s1
    cross join UNNEST(s1.identifier) as i1
    left outer join UNNEST(s1.Diagnosis) as d1 
    where d1 is  null

    ''', limit=1000)
ids2 = [ sub['id'] for sub in res2.result ]
len(ids2)

398

In [100]:
print(len(list(set(ids2)-set(ids1))))
print(len(list(set(ids1)-set(ids2))))
list(set(ids2)-set(ids1))

62
0


['TARGET-10-PAPAJU',
 'TARGET-10-PAPIGH',
 'TARGET-10-PARXLS',
 'GTEX-UTHO-0011-R10A-SM-HAKY2',
 'GTEX-NPJ7-0011-R10A-SM-HAKXW',
 'TARGET-10-PARTPW',
 'TARGET-10-PARUEU',
 'C3L-07035',
 'TARGET-10-PARTBP',
 'TARGET-10-PATKVD',
 'TARGET-10-PAPGMC',
 'TARGET-10-PATGWP',
 'TARGET-10-PASZJW',
 'TARGET-10-PASWFN',
 'TARGET-10-PAPEWV',
 'TARGET-10-PAPDAU',
 'TARGET-10-PAPEZX',
 'TARGET-10-PASJYI',
 'TARGET-10-PASHNK',
 'TARGET-10-PARXMV',
 'TARGET-10-PAPAWY',
 'TARGET-10-PAPDCR',
 'TARGET-10-PAPIGX',
 'TARGET-10-PANYHW',
 'TARGET-10-PANYXC',
 'TARGET-10-PANJRD',
 'TARGET-10-PANRIM',
 'GTEX-P44H-0011-R10A-SM-HAKXX',
 'TARGET-10-PATGBY',
 'TARGET-10-PARPYJ',
 'C3L-07033',
 'TARGET-10-PARSET',
 'TARGET-10-PATFWF',
 'TARGET-10-PANWPI',
 'GTEX-WVLH-0011-R10A-SM-HAKXZ',
 'TARGET-10-PASHUP',
 'C3L-07032',
 'GTEX-RN5K-0011-R10A-SM-HAKXU',
 'TARGET-10-PANKMB',
 'C3L-07034',
 'TARGET-10-PAPGPJ',
 'GTEX-R55F-0011-R10A-SM-HAKXY',
 'TARGET-10-PATGYH',
 'TARGET-10-PATALJ',
 'TARGET-10-PASJJR',
 'TARGET-10

List one of them

In [108]:
res = api_instance.sql_query('v3',
                             '''select p.*
    FROM gdc-bq-sample.cda_mvp.v3 p
    where id = 'TARGET-10-PATCNI' ''',
                      limit=100)
for p in res.result:
    print ('patient {}'.format(p['id']))
    for s in p['ResearchSubject']:
        print ('subject {}'.format(s['id']))
        print('Diagnosis {}'.format(s['Diagnosis']))

patient TARGET-10-PATCNI
subject 9e229e56-f7e1-58f9-984b-a9453be5dc9a
Diagnosis [{'morphology': '9837/3', 'tumor_stage': 'unknown', 'tumor_grade': 'Not Reported', 'Treatment': [], 'id': '63c6c7a2-c3e6-5e2b-9dd6-03bfabef5178', 'primary_diagnosis': 'T lymphoblastic leukemia/lymphoma', 'age_at_diagnosis': 1682}]
subject 7eeced68-1717-4116-bcee-328ac70a9682
Diagnosis []


In [44]:
api_instance.sql_query('v3','''select p.id,  d1.Treatment as treatment
    FROM gdc-bq-sample.cda_mvp.v3 p,
    unnest(ResearchSubject) as s1,
    UNNEST(s1.identifier) as i1,
    UNNEST(s1.Diagnosis) as d1
    where i1.system ='PDC' 
    order by p.id ''')
#print ('result count: {}'.format(len(res.result)))
#print (res)

{'next_url': None,
 'previous_url': None,
 'query_sql': 'select p.id,  d1.Treatment as treatment\n'
              '    FROM gdc-bq-sample.cda_mvp.v3 p,\n'
              '    unnest(ResearchSubject) as s1,\n'
              '    UNNEST(s1.identifier) as i1,\n'
              '    UNNEST(s1.Diagnosis) as d1\n'
              "    where i1.system ='PDC' \n"
              '    order by p.id ',
 'result': [{'id': '01BR001', 'treatment': []},
            {'id': '01BR008', 'treatment': []},
            {'id': '01BR009', 'treatment': []},
            {'id': '01BR010', 'treatment': []},
            {'id': '01BR015', 'treatment': []},
            {'id': '01BR017', 'treatment': []},
            {'id': '01BR018', 'treatment': []},
            {'id': '01BR020', 'treatment': []},
            {'id': '01BR023', 'treatment': []},
            {'id': '01BR025', 'treatment': []},
            {'id': '01BR026', 'treatment': []},
            {'id': '01BR027', 'treatment': []},
            {'id': '01BR030', 'tre

In [112]:
two_id_query = '''select p.id ,  s1.id
    FROM gdc-bq-sample.cda_mvp.v3 p,
    unnest(ResearchSubject) as s1,
    UNNEST(s1.identifier) as i1,
    UNNEST(s1.Diagnosis) as d1
    where i1.system ='GDC' 
    order by p.id '''
api_instance.sql_query('v3',two_id_query,
                      limit=10)

{'next_url': None,
 'previous_url': None,
 'query_sql': 'select p.id ,  s1.id\n'
              '    FROM gdc-bq-sample.cda_mvp.v3 p,\n'
              '    unnest(ResearchSubject) as s1,\n'
              '    UNNEST(s1.identifier) as i1,\n'
              '    UNNEST(s1.Diagnosis) as d1\n'
              "    where i1.system ='GDC' \n"
              '    order by p.id ',
 'result': [{'id': 'be37f1f7-2f98-4f74-bc04-6dd2ae2afcad'},
            {'id': 'e6915db0-7c89-484d-8f9f-15cca68b82fc'},
            {'id': '16614d46-172b-479c-992b-e80a8e9a2c59'},
            {'id': '567fc9e3-17a6-42b1-a896-5e9a9507d1d8'},
            {'id': '54e89878-a1bc-4f5a-9d68-4842a469586e'},
            {'id': 'a1c7b7b9-b8c8-48c3-9420-55497f9318fd'},
            {'id': 'ce3c8b98-e275-4cfd-a379-940d675a564b'},
            {'id': 'e4ce89ef-bcaa-418a-8a6b-3602793b9bbf'},
            {'id': '19d3c861-8a5f-49a2-acc0-b55b25465c35'},
            {'id': 'afae8dce-294a-4108-bb28-376f804ae5c4'}]}

In [113]:
import requests
def runAPIQuery(querystring, limit=None):
    cdaURL = 'https://cda.cda-dev.broadinstitute.org/api/v1/sql-query/v3'
    #Using a limit:
    if limit is not None:
        cdaURL = "{}?limit={}".format(cdaURL, str(limit))
        
    headers = {'accept' : 'application/json', 'Content-Type' : 'text/plain'}

    request = requests.post(cdaURL, headers = headers, data = querystring)

    if request.status_code == 200:
        return request.json()
    else:
        raise Exception ("Query failed code {}. {}".format(request.status_code,query))

{'query_sql': "select p.id ,  s1.id\n    FROM gdc-bq-sample.cda_mvp.v3 p,\n    unnest(ResearchSubject) as s1,\n    UNNEST(s1.identifier) as i1,\n    UNNEST(s1.Diagnosis) as d1\n    where i1.system ='GDC' \n    order by p.id ",
 'result': [{'id': 'be37f1f7-2f98-4f74-bc04-6dd2ae2afcad'},
  {'id': 'e6915db0-7c89-484d-8f9f-15cca68b82fc'},
  {'id': '16614d46-172b-479c-992b-e80a8e9a2c59'},
  {'id': '567fc9e3-17a6-42b1-a896-5e9a9507d1d8'},
  {'id': '54e89878-a1bc-4f5a-9d68-4842a469586e'},
  {'id': 'a1c7b7b9-b8c8-48c3-9420-55497f9318fd'},
  {'id': 'ce3c8b98-e275-4cfd-a379-940d675a564b'},
  {'id': 'e4ce89ef-bcaa-418a-8a6b-3602793b9bbf'},
  {'id': '19d3c861-8a5f-49a2-acc0-b55b25465c35'},
  {'id': 'afae8dce-294a-4108-bb28-376f804ae5c4'}],
 'previous_url': None,
 'next_url': None}